### DISCUSSION TITLE
clickhouse udf DB::ErrnoException: Cannot write into pipe, errno: 32, strerror: Broken pipe: While executing JSONEachRowRowOutputFormat

### DISCUSSION THREAD

**Discussion by QiMingChina** [ClickHouse/ClickHouse#53197] [Q&A]
I use Python to write clickhouse udf
clickhouse version 23.3.2 Python 3.9.12
I added the configuration <user_defined_executable_functions_config>*_function.xml</user_defined_executable_functions_config> in the config.xml file
I added the configuration file  getusername_function.xml and added the clickhouse user permission in the/etc/clickhouse server path
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/a8d47f35-b8e9-4ad0-ab47-418a9856b292)
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/5f68a370-8286-46fb-8cda-63c6bc9c2250)
```
<functions>
    <function>
        <type>executable_pool</type>
        <name>username_function</name>
        <return_type>String</return_type>
        <return_name>result_username</return_name>
        <argument>
            <type>String</type>
            <name>first_name</name>
        </argument>
        <argument>
            <type>String</type>
            <name>second_name</name>
        </argument>
        <argument>
            <type>String</type>
            <name>hobby</name>
        </argument>
        <format>JSONEachRow</format>
        <command>username.py</command>
    </function>
</functions>
```
I  place Python file a in user_scripts path 
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/268c8767-af56-4bae-8b54-b4fe7e0b1cbd)
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/595f273b-7656-4321-9ac9-47505d45385e)
```
#!/servers/software/anaconda3/bin/python

import sys
import json
import logging

logging.basicConfig()
logging.getLogger().setLevel(logging.INFO)
file_handler = logging.FileHandler('/data/clickHousedata/user_scripts/username-log.txt')
logging.getLogger().addHandler(file_handler)


def username(first_name, second_name, hobby):
    user_info = '姓名：{0}  爱好：{1}'.format(first_name + second_name, hobby)
    return user_info


if __name__ == '__main__':
    for line in sys.stdin:
        logging.info('输入参数列表：{0}'.format(line))
        arg_dict = json.loads(line)
        first_name = arg_dict['first_name']
        second_name = arg_dict['second_name']
        hobby = arg_dict['hobby']
        info=username(first_name,second_name,hobby)
        result = {'result_username':info}
        print(json.dumps(result), end='\n')
        sys.stdout.flush()
```

reload function
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/6d12cd48-a513-4a6d-9851-35451f15d526)

raise error
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/b22e83e0-ccd6-4b86-8be0-cdac9e1b58e5)

**Comment by mwoenker** [comment]
Are you able to run the script manually? For example if you run

```
# from your user_scripts directory
echo '{ "first_name": "first", "second_name": "second", "hobby": "hobby" }' | ./username.py
```

do you get correct output?

**Reply by QiMingChina** [reply]
I ran the script and found a file format issue
```
bash: ./username.py: /servers/software/anaconda3/bin/python^M: bad interpreter: No such file or directory
```
After solving this problem, the following problems occurred
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/fe84c669-8834-4ee2-bf55-c530a5e387e7)

I have reviewed this issue but it has not been resolved
[UDF in shell: expected 0 row(s)](https://github.com/ClickHouse/ClickHouse/issues/41611)

**Comment by RemilYoucef** [comment]
@QiMingChina 
The error you are getting indicates that there is an issue with the interpreter specified in the first line of your script which is 
#!/usr/bin/python3 if you run the script from the directory using echo "input" | your_script.py


The same problems happens to me, becauses I uploaded directly a python file created under the windows system. 

The presence of ^M characters in the error message is because your file has been created or modified on a system that uses different line endings, and these characters are hidden when you read your file either with cat or vim. 
What you can do, is executing this: 
dos2unix your_script.py 
or this: 
sed -i 's/\r$//' your_script.py
